### PR TITLE
Extract reusable SequenceCoverage API and add unit test

### LINKED ISSUE #6240: Implement a getCoverage(String, vector<AASequence/String/PeptideHit>) and expose in pyopenms
This is a common task that people need to do in python.
There can be multiple ways of implementing it.

### PR DESCRIPTION
**Pull request by Srikanthkn0**
## Description

This PR extracts the sequence coverage calculation logic from
`SequenceCoverageCalculator` into a reusable CHEMISTRY-level API
(`OpenMS::SequenceCoverage`).

The new API provides a clean `getCoverage` function operating on
`AASequence` inputs and can be reused by other tools and language bindings.
A dedicated unit test (`SequenceCoverage_test`) has been added to validate
the correctness of the implementation.

No existing tool behavior is changed by this PR.

Fixes #6240

## Summary by CodeRabbit

* **New Features**
  * Public sequence-coverage calculation added to CHEMISTRY to compute protein coverage percentages from peptide sets (supports overlapping matches; returns 0–100).
  * Python bindings expose the sequence-coverage functionality for scripting and analysis.
  * Sequence coverage calculator tool updated to use the new utility.

* **Tests**
  * New unit tests covering typical cases, overlaps, and edge cases (empty protein/peptides).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/openms/include/OpenMS/CHEMISTRY/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+set(CHEMISTRY_SOURCES
+AASequunce.cpp
+EmpericalFormula.cpp
+SequenceCoverage.cpp
```
[the remark this answers is not in the document]
Comment (Srikanthkn0) [reply src/openms/include/OpenMS/CHEMISTRY/CMakeLists.txt]: Thanks for catching this — you’re absolutely right.
The CMakeLists.txt under include/OpenMS/CHEMISTRY was incorrect and has been removed.
The build now relies solely on the existing sources.cmake files, consistent with the rest of OpenMS.

File: src/openms/include/OpenMS/CHEMISTRY/SequenceCoverage.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+
+  class OPENMS_DLLAPI SequenceCoverage
+  {
+  public:
+ 
+    static double getCoverage(
+      const AASequence& protein,
+      const std::vector<AASequence>& peptides
+    );
```
[the remark this answers is not in the document]
Comment (Srikanthkn0) [reply src/openms/include/OpenMS/CHEMISTRY/SequenceCoverage.h:37]: I’ve added pyOpenMS bindings by introducing SequenceCoverage.pxd and ensured the public API is fully documented with Doxygen comments.

**Review by timosachsenberg — APPROVED** [review]
Looks correct to me. Can you fix the some minor style issues like bracket placement and  equality sign?

**Comment by Srikanthkn0** [comment]
Thanks for the review!
Sure, I’ll fix the minor style issues you mentioned (bracket placement and equality usage) and update the PR.

**Review by timosachsenberg — CHANGES_REQUESTED** [review]
Please also make sure that this is actually used. You now just copied the functionality used somewhere else

File: src/pyOpenMS/pxds/SequenceCoverage.pxd [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+from pyopenms cimport AASequence
+
+cdef extern from "<OpenMS/CHEMISTRY/SequenceCoverage.h>" namespace "OpenMS":
+
+    cdef cppclass SequenceCoverage:
+        @staticmethod
+        double getCoverage(
+            const AASequence& protein,
+            const vector[AASequence]& peptides
+        ) except +
```
Comment (Srikanthkn0) [inline src/pyOpenMS/pxds/SequenceCoverage.pxd]: Thanks for the review. I’ve added brief documentation comments to the pyOpenMS .pxd file (both above the extern block and the getCoverage method) to meet the documentation guidelines, and pushed the update.

File: src/pyOpenMS/pxds/SequenceCoverage.pxd [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    cdef cppclass SequenceCoverage:
+        SequenceCoverage() except + nogil
+        SequenceCoverage(SequenceCoverage&) except + nogil
+        # Compute the percentage sequence coverage of a protein by given peptides.
+        @staticmethod
+        double getCoverage(
+            const AASequence& protein,
+            const vector[AASequence]& peptides
+        ) except +
```
Comment (timosachsenberg) [inline src/pyOpenMS/pxds/SequenceCoverage.pxd]:
```suggestion
        ) except + nogil 
```

Comment (timosachsenberg) [reply src/pyOpenMS/pxds/SequenceCoverage.pxd]: and proper wrap-doc

File: src/topp/SequenceCoverageCalculator.cpp [hunk 5]
```diff
@@ -133,8 +106,8 @@ class TOPPSequenceCoverageCalculator :
     os << "proteinID\tcoverage (%)\tunique hits\n";
     for (Size j = 0; j < proteins.size(); ++j)
     {
-      coverage.clear();
-      coverage.resize(proteins[j].sequence.size(), 0);
+        AASequence protein_seq = AASequence::fromString(proteins[j].sequence);
+        std::vector<AASequence> peptides;
```
Comment (timosachsenberg) [inline src/topp/SequenceCoverageCalculator.cpp]: fix indentation

File: src/topp/SequenceCoverageCalculator.cpp [hunk 6]
```diff
@@ -155,16 +128,7 @@ class TOPPSequenceCoverageCalculator :
...
-            getStartAndEndIndex(proteins[j].sequence, temp_hits[0].getSequence().toUnmodifiedString(), indices);
-            for (Size k = indices.first; k < indices.second; ++k)
-            {
-              coverage[k] = 1;
-            }
-            if (indices.first != indices.second)
-            {
-              // os <<  temp_hits[0].getSequence().toUnmodifiedString() << endl;
-            }
+              peptides.push_back(temp_hits[0].getSequence());
```
Comment (timosachsenberg) [inline src/topp/SequenceCoverageCalculator.cpp]: same here

Comment (timosachsenberg) [reply src/topp/SequenceCoverageCalculator.cpp]: and below as well

**Comment by timosachsenberg** [comment]
+ add pyopenms test

**Comment by timosachsenberg** [comment]
in src/topp/SequenceCoverageCalculator.cpp:177 replace the current mixed-key guard+insert with something like:
```
  const AASequence& seq = temp_hits[0].getSequence();
  if (seq.isModified())
  {
    temp_modified_unique_peptides.emplace(seq.toString(), 0);
  }
```
I think the original implementation is not correct there.

**Comment by Srikanthkn0** [comment]
@timosachsenberg Thanks for the review!

I’ve addressed all the points:
	•	Added nogil and improved wrap-documentation in the pyOpenMS .pxd bindings
	•	Added a dedicated pyOpenMS unit test for SequenceCoverage
	•	Fixed indentation issues in SequenceCoverageCalculator.cpp
	•	Replaced the mixed-key modified-peptide handling with an AASequence::isModified()-based check as suggested

Please let me know if anything else should be adjusted.

File: src/pyOpenMS/tests/unittests/test_SequenceCoverage.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class TestSequenceCoverage(unittest.TestCase):
+
+    def test_partial_coverage(self):
+        """Partial sequence coverage is computed correctly"""
+        protein = oms.AASequence.fromString("PEPTIDEAAAAAA")
+        peptides = [
+            oms.AASequence.fromString("PEPTIDE")
+        ]
```
Comment (timosachsenberg) [inline src/pyOpenMS/tests/unittests/test_SequenceCoverage.py]: This test is good. Keep it.  But also test.
```
        protein = oms.AASequence.fromString("PEPTIDEPEPTIDE")
        peptides = [
            oms.AASequence.fromString("PEPTIDE")
        ]
```
as it tells us how the algorithm handles peptides that map multiple times

Comment (Srikanthkn0) [reply src/pyOpenMS/tests/unittests/test_SequenceCoverage.py]: Thanks for the suggestion. I’ve kept the non-repeating case and added an additional test using PEPTIDEPEPTIDE to explicitly cover the scenario where a peptide maps multiple times within the same protein sequence.

**Comment by timosachsenberg** [comment]
Did you Build and run pyopenms teste?

**Comment by Srikanthkn0** [comment]
@timosachsenberg I built OpenMS and ran the C++ tests locally, and everything passed on my side.
Unfortunately I wasn’t able to run the pyOpenMS tests locally due to missing system dependencies on my setup, but the added pyOpenMS test follows existing patterns and should be exercised by CI. Please let me know if you’d like me to try a specific setup or adjust anything.

**Comment by timosachsenberg** [comment]
Running test is mandatory. Please reopen a clean PR once you fulfill the minimum requirements
